crates/graphql_artifact_generation/src/generate_artifacts.rs

```diff
@@ -614,6 +614,114 @@ fn format_type_for_js_inner(
     }
 }
 
+pub(crate) fn generate_parameters<'a>(
+    schema: &ValidatedSchema,
+    argument_definitions: impl Iterator<Item = &'a VariableDefinition<SelectableServerFieldId>>,
+    indentation_level: u8,
+) -> String {
+    let mut parameters_type = "{\n".to_string();
+    for arg in argument_definitions {
+        let is_optional = !matches!(arg.type_, GraphQLTypeAnnotation::NonNull(_));
+        parameters_type.push_str(&format!(
+            "{}readonly {}{}: {},\n",
+            "  ".repeat((indentation_level + 1) as usize),
+            arg.name.item,
+            if is_optional { "?" } else { "" },
+            format_parameter_type(schema, arg.type_.clone(), indentation_level + 1)
+        ));
+    }
+    parameters_type.push_str(&format!("{}}};", "  ".repeat(indentation_level as usize)));
+    parameters_type
+}
+
+fn format_parameter_type(
+    schema: &ValidatedSchema,
+    type_: GraphQLTypeAnnotation<SelectableServerFieldId>,
+    indentation_level: u8,
+) -> String {
+    match type_ {
+        GraphQLTypeAnnotation::Named(named_inner_type) => {
+            format!(
+                "{} | null | void",
+                format_server_field_type(schema, named_inner_type.item, indentation_level)
+            )
+        }
+        GraphQLTypeAnnotation::List(list) => {
+            format!(
+                "ReadonlyArray<{}> | null",
+                format_server_field_type(schema, *list.0.inner(), indentation_level)
+            )
+        }
+        GraphQLTypeAnnotation::NonNull(non_null) => match *non_null {
+            NonNullTypeAnnotation::Named(named_inner_type) => {
+                format_server_field_type(schema, named_inner_type.0.item, indentation_level)
+            }
+            NonNullTypeAnnotation::List(list) => {
+                format!(
+                    "ReadonlyArray<{}>",
+                    format_server_field_type(schema, *list.0.inner(), indentation_level)
+                )
+            }
+        },
+    }
+}
+
+fn format_server_field_type(
+    schema: &ValidatedSchema,
+    field: SelectableServerFieldId,
+    indentation_level: u8,
+) -> String {
+    match field {
+        SelectableServerFieldId::Object(object_id) => {
+            let mut parameters_type = "{\n".to_string();
+            for (name, field_definition) in schema
+                .server_field_data
+                .object(object_id)
+                .encountered_fields
```

Comment:
I think a good solution would be to make `SchemaServerField` an enum with variants `Discriminator`, `StrongId`, and `EverythingElse`. But, in the short term, I think we should make a separate PR that:

* adds a field to `SchemaServerField`
```
// TODO remove this. This is indicative of poor modeling.
is_discriminator: bool
```
* set that to true only in [`process_type_definition:691`](https://github.com/isographlabs/isograph/blob/0254e4b5afdec037715e443012b9e32b408645fc/crates/isograph_schema/src/process_type_definition.rs#L691-L698)
* add a method to to `SchemaObject` that is called `non_discriminator_fields`. Also, add a `// TODO remove this, this is indicative of poor modeling` comment 😂 
* call that here

Reply:
#150 for an actual solution, which needs some thinking